packages/caver-account/src/accountKey/accountKeyRoleBased.js

```diff
@@ -0,0 +1,177 @@
+        if (
+            accountKey !== undefined &&
+            !(accountKey instanceof AccountKeyLegacy) &&
+            !(accountKey instanceof AccountKeyPublic) &&
+            !(accountKey instanceof AccountKeyFail) &&
+            !(accountKey instanceof AccountKeyWeightedMultiSig)
+        ) {
+            return false
+        }
+    }
+    return true
+}
+
+/**
+ * Representing an AccountKeyRoleBased.
+ * @class
+ */
+class AccountKeyRoleBased {
+    /**
+     * Decodes an RLP-encoded AccountKeyRoleBased string.
+     * @param {string} rlpEncodedKey - An RLP-encoded AccountKeyRoleBased string.
+     * @return {AccountKeyRoleBased}
+     */
+    static decode(rlpEncodedKey) {
+        rlpEncodedKey = utils.addHexPrefix(rlpEncodedKey)
+        if (!rlpEncodedKey.startsWith(ACCOUNT_KEY_TAG.ACCOUNT_KEY_ROLE_BASED_TAG))
+            throw new Error(
+                `Cannot decode to AccountKeyRoleBased. The prefix must be ${ACCOUNT_KEY_TAG.ACCOUNT_KEY_ROLE_BASED_TAG}: ${rlpEncodedKey}`
+            )
+
+        const keys = RLP.decode(`0x${rlpEncodedKey.slice(ACCOUNT_KEY_TAG.ACCOUNT_KEY_ROLE_BASED_TAG.length)}`)
+        const accountKeys = []
+        for (const key of keys) {
+            if (key.startsWith(ACCOUNT_KEY_TAG.ACCOUNT_KEY_ROLE_BASED_TAG)) throw new Error('Nested role based key.')
+            if (key.startsWith(ACCOUNT_KEY_TAG.ACCOUNT_KEY_NIL_TAG)) {
+                accountKeys.push(undefined)
+            } else if (key.startsWith(ACCOUNT_KEY_TAG.ACCOUNT_KEY_LEGACY_TAG)) {
+                accountKeys.push(AccountKeyLegacy.decode(key))
+            } else if (key.startsWith(ACCOUNT_KEY_TAG.ACCOUNT_KEY_PUBLIC_TAG)) {
+                accountKeys.push(AccountKeyPublic.decode(key))
+            } else if (key.startsWith(ACCOUNT_KEY_TAG.ACCOUNT_KEY_FAIL_TAG)) {
+                accountKeys.push(AccountKeyFail.decode(key))
+            } else if (key.startsWith(ACCOUNT_KEY_TAG.ACCOUNT_KEY_WEIGHTED_MULTISIG_TAG)) {
+                accountKeys.push(AccountKeyWeightedMultiSig.decode(key))
+            } else {
+                throw new Error(`Failed to decode RLP-encoded account key. Invalid RLP-encoded account key ${key}`)
+            }
+        }
+        return new AccountKeyRoleBased(accountKeys)
+    }
+
+    /**
+     * Creates an instance of AccountKeyRoleBased.
+     * @param {Array.<AccountKeyLegacy|AccountKeyFail|Array.<string>>} roleBasedPubArray - An array of public key strings.
+     * @param {Array.<object>} options - An array of options which defines threshold and weight.
+     * @return {AccountKeyRoleBased}
+     */
+    static fromRoleBasedPublicKeysAndOptions(roleBasedPubArray, options = [{}, {}, {}]) {
+        const accountKeys = []
+        // Format will be like below
+        // keyArray = [[pub, pub], [pub], [pub, pub, pub]]
+        // keyArray = [accountKeyLegacy, accountKeyFail, [pub, pub, pub]]
+        // keyArray = [accountKeyLegacy, accountKeyFail, [pub]]
```

Comment:
What about this? Do we support it?
public key not surrounded by []?
`keyArray = [accountKeyLegacy, accountKeyFail, pub]`

Reply:
My intention was originally support non-array format only AccountKeyLegacy, AccountKeyFail.

Because if AccountKeyRoleBased, and each role uses AccountKeyPublic, if each role is not defined as an array, it is confusing whether to judge it as AccountKeyWeightedMultiSig or AccountKeyRoleBased with AccountKeyPublic in the case below.
[pub, pub, pub]

Reply:
Hmm... I'm not sure then, we need to support non-array format for AccountKeyLegacy, AccountKeyFail. To me, it is little bit confusing that both accountKeyLegacy and [accountKeyLegacy] are supported, but pub is not supported while [pub] is supported.

Reply:
Oh, okay. i know what your thinking! I think that is more clear. :) I will edit!

Reply:
I edited :) !!